File: day11/three_sum.cpp
#include<vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums){
        std::vector<std::vector<int>> result;

        std::sort(nums.begin(), nums.end());

        int numSize = nums.size();
        int prevKey = INT_MAX;

        for(int i = 0; i < numSize; ++i){        
            int l = i+1;
            int r = nums.size()-1;
            if(prevKey == nums[i]){
                continue;
            }

            while(r > l){
                if(nums[i] + nums[l] + nums[r] > 0){
                    --r;
                }
                else if (nums[i] + nums[l] + nums[r] < 0){
                    ++l;
                }
                else{
                    result.push_back({nums[i],nums[l],nums[r]});
                    ++l;
                    while(nums[l] == nums[l-1] && l < r){
                        ++l;
                    }
                }
            }
            prevKey = nums[i];
        }
        return result;
    }
};

File: day_12/three_sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        nums.sort()

        for i, j in enumerate(nums):
            if i > 0 and j == nums[i - 1]:
                continue
            l = i + 1
            r = len(nums) - 1
            while l < r:
                three_sum = j + nums[l] + nums[r]
                if three_sum > 0:
                    r -= 1
                elif three_sum < 0:
                    l += 1
                else:
                    result.append([j, nums[r], nums[l]])
                    l += 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
        return result
